### PR TITLE
#163518683 User can get email verification

### PR DESCRIPTION
**Pull request by niomwungeri-fabrice**
#### What does this PR do?
Email verification 
#### Description of Task to be completed?
After a user has been created successfully the platform sends the email via email,
the user is able to click on the link in order to get access to other services like creating an article.
The is an option of resending the email if the email wasn's sent on time.
EndPoints:
* `POST api/v1/users` for creating an account to the system, and receive a verification email.
* `GET api/v1/users/confirm/:auth_token` for email verification.
* `PUT api/v1/users/resend` for resending email verification.
#### How should this be manually tested?
* Fetch the branch `ft-user-get-verification-link-163518683`
* Checkout the branch `ft-user-get-verification-link-163518683`
* Setup your database. 
* Run tests `yarn run test`
* Start the server `yarn run server`
* Create a user `POST: /api/v1/users` with `email, password, and username`
* Go to your email click on the `verify my email` which was sent while registering it will trigger this endPoint `GET api/v1/users/confirm/:auth_token` using Postman client. 
* You should see a User verified successfully. 
* If the email didn't come you could use this endpoint to resend the email `PUT api/v1/users/resend` you only provide email as follow:
`
{
	"email":"example@gmail.com"
}
`
#### What are the relevant pivotal tracker stories?
163518683

### CODE REVIEW

**Review by YvesIraguha — APPROVED** [review]

File: controllers/usersController.js [hunk 1]
```diff
@@ -20,9 +21,9 @@ class UserController {
    *
    * ResetPassword.
    *
```
Comment (puritybirir) [inline controllers/usersController.js]: Add description to comments

File: controllers/usersController.js [hunk 2]
```diff
@@ -112,6 +113,80 @@ class UserController {
         .json({ message: "Password update failed", errors: error.stack });
     }
   }
+
```
Comment (puritybirir) [inline controllers/usersController.js:138]: JSDoc comments

File: controllers/usersController.js [hunk 3]
```diff
@@ -112,6 +113,80 @@ class UserController {
...
+  static async resendVerificationEmail(req, res) {
+    try {
+      const { email } = req.body;
+      if (!email) return res.status(404).json({ message: "Email is required" });
```
Comment (puritybirir) [inline controllers/usersController.js]: Incorrect status code. 404 is for not found

Comment (puritybirir) [reply controllers/usersController.js]: Invalid data should be a 400

File: controllers/usersController.js [hunk 4]
```diff
@@ -112,6 +113,80 @@ class UserController {
...
+      if (!isEmailValid(email))
+        return res.status(404).json({ message: "Invalid email" });
```
Comment (puritybirir) [inline controllers/usersController.js]: use constants file for the status code

File: controllers/usersController.js [hunk 5]
```diff
@@ -112,6 +113,80 @@ class UserController {
...
+        where: { email }
+      });
+      if (!user) {
+        return res.status(404).json({
+          message: "User not found"
+        });
+      }
+      if (user.dataValues.isVerified) {
+        return res.status(200).json({
+          message: "User already verified!"
```
Comment (puritybirir) [inline controllers/usersController.js]: Message should be : User verified

File: controllers/usersController.js [hunk 6]
```diff
@@ -112,6 +113,80 @@ class UserController {
...
+      return res.status(200).json({
+        message: "Email sent successufully"
+      });
+    } catch (error) {
+      return res.status(500).json({
+        messege: error
+      });
+    }
+  }
+
```
Comment (puritybirir) [inline controllers/usersController.js:183]: JSDoc comments

File: migrations/20190219113957-add-isVerified-to-user.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    /*
+      Add altering commands here.
```
Comment (puritybirir) [inline migrations/20190219113957-add-isVerified-to-user.js]: Remove comments

File: tests/integration/emailConfirmation.test.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      .request(app)
+      .get(`/api/v1/users/confirm/${userToken}`);
+    expect(results.status).equal(200);
+    expect(results.body.message).equal("Email confirmed successfully!");
+  });
+  it("/PUT it should fail with wrong token", async () => {
+    const results = await chai
+      .request(app)
+      .get(`/api/v1/users/confirm/${wrongToken}`);
+    expect(results.status).equal(404);
```
Comment (puritybirir) [inline tests/integration/emailConfirmation.test.js]: Use correct status codes

**Review by puritybirir — CHANGES_REQUESTED** [review]

File: controllers/usersController.js [hunk 9]
```diff
@@ -53,17 +66,21 @@ class UserController {
...
-            .status(500)
+            .status(INTERNAL_SERVER_ERROR)
             .json({ message: "Error while sending email", data: emailResponse });
         }
       });
     } catch (error) {
-      res.status(500).json({ message: "Sending email failed", errors: error.stack });
+      res
+        .status(INTERNAL_SERVER_ERROR)
+        .json({ message: "Sending email failed", errors: error.stack });
```
Comment (espoirMur) [inline controllers/usersController.js]: Why sending error.stack to the user??

File: controllers/usersController.js [hunk 10]
```diff
@@ -104,10 +123,103 @@ class UserController {
       });
     } catch (error) {
       res
-        .status(500)
+        .status(INTERNAL_SERVER_ERROR)
```
Comment (espoirMur) [inline controllers/usersController.js:126]: Same thing here, why sending errors.stack to the user?

File: tests/helpers/testData.js [hunk 11]
```diff
@@ -20,9 +20,21 @@ const users = {
     email: "fake.email@gmail.com",
     password: "password",
     username: "jean786786"
+  },
+  dummyUser3: {
+    email: "fabrice.niyomwungeri@andela.com",
```
Comment (espoirMur) [inline tests/helpers/testData.js:25]: don't share your email on GitHub please, you may get spammed by hacker...

File: tests/integration/comment.test.js [hunk 12]
```diff
@@ -32,6 +33,8 @@ before(done => {
                 .set("Authorization", `Bearer ${token}`)
                 .send(newArticle)
                 .end((err, res) => {
+                  console.log(res.body);
+                  console.log(res.status);
```
[the remark this answers is not in the document]
Comment (espoirMur) [reply tests/integration/comment.test.js]: Thanks, @houndci-bot !! Good Job.

**Review by espoirMur — CHANGES_REQUESTED** [review]

**Review by puritybirir — APPROVED** [review]

**Review by espoirMur — APPROVED** [review]

**Pull request merged**